cdap-master/src/main/java/io/cdap/cdap/master/upgrade/UpgradeJobMain.java

```diff
@@ -0,0 +1,120 @@
+  private static final String SCHEDULED = "SCHEDULED";
+
+  public static void main(String[] args) {
+    if (args.length != 1) {
+      throw new RuntimeException(
+        String.format("Invalid number of arguments to UpgradeJobMain. Needed 1, found %d", args.length));
+    }
+    ConnectionConfig connectionConfig = ConnectionConfig.builder()
+      .setHostname("cdap-" + args[0] + "-router")
+      .setPort(PORT)
+      .setSSLEnabled(false)
+      .build();
+    ClientConfig clientConfig =
+      ClientConfig.builder()
+        .setDefaultReadTimeout(DEFAULT_READ_TIMEOUT_MILLIS)
+        .setConnectionConfig(connectionConfig)
+        .build();
+    ApplicationClient applicationClient = new ApplicationClient(clientConfig);
+    ScheduleClient scheduleClient = new ScheduleClient(clientConfig);
+    ProgramClient programClient = new ProgramClient(clientConfig);
+    NamespaceClient namespaceClient = new NamespaceClient(clientConfig);
+    RetryStrategy retryStrategy =
+      RetryStrategies.timeLimit(30, TimeUnit.SECONDS,
+                                RetryStrategies.exponentialDelay(10, 500, TimeUnit.MILLISECONDS));
+
+    try {
+      Retries.supplyWithRetries(
+        () -> {
+          try {
+            for (NamespaceMeta namespaceMeta : namespaceClient.list()) {
+              NamespaceId namespaceId = namespaceMeta.getNamespaceId();
+              for (ApplicationRecord record : applicationClient.list(namespaceId)) {
+                ApplicationId applicationId =
+                  new ApplicationId(namespaceId.getNamespace(), record.getName(), record.getAppVersion());
+                List<WorkflowId> workflowIds =
+                  applicationClient.get(applicationId).getPrograms().stream()
+                    .filter(programRecord -> programRecord.getType().equals(ProgramType.WORKFLOW))
+                    .map(programRecord -> new WorkflowId(applicationId, programRecord.getName()))
+                    .collect(Collectors.toList());
+                for (WorkflowId workflowId : workflowIds) {
+                  List<ScheduleId> scheduleIds =
+                    scheduleClient.listSchedules(workflowId).stream()
+                      .map(scheduleDetail ->
+                             new ScheduleId(namespaceId.getNamespace(), record.getName(),
+                                            record.getAppVersion(), scheduleDetail.getName()))
+                      .collect(Collectors.toList());
+                  for (ScheduleId scheduleId : scheduleIds) {
+                    if (scheduleClient.getStatus(scheduleId).equals(SCHEDULED)) {
+                      scheduleClient.suspend(scheduleId);
+                    }
+                  }
+                }
+              }
+
+              // All schedules are stopped, now stop all pipelines
+              programClient.stopAll(namespaceId);
+            }
+
+            return null;
+          } catch (IOException | NotFoundException e) {
```

Comment:
Instead of wrapping everything in a try-catch, you can use callWithRetries(), which allows throwing an exception and supports taking a predicate that defines which exceptions are retryable.